### DISCUSSION TITLE
Core Principles and Features List

### DISCUSSION THREAD

**Discussion by wai-lin** [mm-ninja-turtles/turtle-express#7] [Ideas]
# Core Principles

I think a framework should provide a full experience for a full featured backend applications.
There're some features that a backend framework should provide if possible. Also this should not be deviated from the express's core principles so that framework users can extend and customize with their familiar apis.

> NOTE:
>
> If we decide to support a feature, it must be good enough for the end users to not think about it
> and it must not deviate from the core principles of the framework.
___

## Features List

These are the features that we might implement as a core feature of the frameworks.

**Symbols Reference**
- 🚧 `:construction:` working on it
- 👍 `:+1:` added to todo list
- 👎 `:-1:` no plan of doing for the moment
- ✅  `:white_check_mark:` feature implemented
- ❓ `:question:` still in discussion

- [ ] **Typescript** 🚧
  - [x] end to end type safety support for `resolver` function ✅ 
    - [ ] auto generate REST client with type safety api client 👍

- [ ] **Technologies**
  - [x] `REST` support ✅  
  - [ ] `GraphQL` support
  - [ ] `gRPC` support
  - [ ] `Redis` support

- [x] **Auto Pagination Calculator Support** ✅
  - [x] Pagination can be calculated with `total-count`, `limit-count` and `offset` numbers (it would be convenient if there's a helper function like `paginate({ totalCount: 213, limit: 10, offset: 0 })` and the function returns necessary data like `nextOffset`, `pagesLeft`, `currentPage` etc.)

- [ ] **Web Standard APIs** 👎 

- [ ] **Plugin API** ❓

- [x] **Fetch API Support** ✅
  - Until the latest node v16 flag the `fetch` api as stable release and as well as to support node version lower than v16.15.0, will bind the `fetch` api as pollyfill with `cross-fetch` or `node-fetch` libraries.

- [ ] **Database** 👎
  - [ ] `MySQL`
  - [ ] `MariaDB`
  - [ ] `PostgreSQL`
  - [ ] `SQLite`
  - [ ] `Firestore`
  - [ ] `OracleDB`
  - [ ] `DynamoDB`
  - [ ] `MongoDB`
  - [ ] `Neo4J`
  - [ ] Support for popular DB ORM libraries
    - [ ] `Prisma`
    - [ ] `TypeORM`
    - [ ] `Knex`
    - [ ] `Objection.JS`
    - [ ] `Mongoose`
    - [ ] `Mongo`

- [ ] **Debugging** 👍
  - [ ] request, response logging with default to saving in a `logs` folder
  - [ ] `logs` should be browsable from a specific end point (with syntax highlighting preview) `eg. /logs`
  - [ ] end user should be able to customize what they want to log and where to log `eg. database`

- [ ] **Security** 👎
  - [ ] should provide default good enough security practices config
  - [ ] Authentication
    - [ ] build in `passport` support
  - [ ] Authorization
    - [ ] `role` based authorization
  - [ ] `csrf` protection
  - [ ] rate limiting
  - [ ]  cors
  - [ ]  encryption and hashing

- [ ] **Public facing API Documentation** 🚧
  - [ ] for `REST`, `swagger ui` 🚧 
  - [ ] `GraphQL`

- [ ] **Documentation** 🚧
  - [ ] the documentation must be good enough for the users to just read and code (don't want them to google when they want to do something)
  - [ ] example rich documentation
  - [ ] detailed explained `API References`

- [ ] **Web Technologies** 👎
  - [ ] `File Upload`
  - [ ] `Resource Optimization (minify images, etc)`
  - [ ] `WebSockets`
  - [ ] `Queues`
  - [ ] `Schedueling (cron)`
  - [ ] `Streaming`
  - [ ] `Email`
  - [ ] `OTP`
  - [ ] `Push Notification`
  - [ ] `MicroServices`

**Comment by wai-lin** [comment]
# Web Standard APIs Support?

@mm-ninja-turtles/core-maintainers 

I think whether it would be a good decision or not to support Web Standard APIs. Let me know your thoughts and concerns. Following are my opinions.

Currently `express` has it's own `request` and `response` object API like:

```ts
res.status(200).json({ msg: 'Hello' })
```

Equivalent to this in Web Standard would be like:

```ts
const body = JSON.stringify({ msg: 'Hello' })
const response = new Response(body, { status: 200 })
```

___

## Benefits

1. Framework users will be more familiar with `Web Standard APIs` as they use the framework.
1. Backend and Frontend code base will be more similar. `eg. remix`
1. In case if we want to migrate from `express` to other framework, our API will not change. We only need to create a new `adapter` package for that framework.

## Downside

1. Need to develop `adapter layer` between `express's Api` and `Web Standard Api`.
1. There might be some edge cases with the `express` compatibility. (it might be doable since there are other frameworks also doing this)

**Reply by wai-lin** [reply]
After some consideration, I decided that using web standard api should only be an opt-in feature since our goal is to provide awesome `express` experience without deviating from the `express` as much as possible. So, if we decide to develop an adapter, it should only be as a pluggable extension and web standard api will add to the `context` of the handler function instead of completely changing the whole `request` and `response` object provided by `express`.

**Reply by PyaePhyoKyaw61194** [reply]
agree with that @wai-lin

**Comment by PyaePhyoKyaw61194** [comment]
I found that _path_ method of our _router_ accepts two arguments type (_Path,PathOptions_). Where will that _PathOptions_ be used? anything equivalent in express?

**Reply by wai-lin** [reply]
You can find it's potential usage in [api proposal](https://github.com/mm-ninja-turtles/bulbasaur-express/discussions/2) like api router versioning and maybe other possibilities.

**Comment by PyaePhyoKyaw61194** [comment]
The return type _PathReturnType_ of _path_ method also contains __handlerColl_ and __path_. We may only be needed them for internal use. Any ideas about hiding them from framework users?

**Reply by wai-lin** [reply]
currently I thinks it's the only way. `zod` also use this pattern.

**Reply by wai-lin** [reply]
otherwise we'll need state management library and it'll complicate things more.

**Comment by PyaePhyoKyaw61194** [comment]
Another scenario I think is, what should happen if the framework user doesn't define the necessary validation checking (here is zod) in _response_ object of the _handler_? For example : 
```
hello.handler({
  method: 'get',
  response: {
    100: z.string({}),
  },
  resolver() {
    return { 200: 'Hello' }
  },
})

```

**Reply by wai-lin** [reply]
Haven't test it yet. Maybe nothing should response since the framework will only response what is defined inside response object schema.
